### ISSUE TITLE
`weakPointer`: how should the type be represented in the language?

### ISSUE THREAD

**Issue by jeremiah-corrado** [chapel-lang/chapel#20956]
The [prototype implementation](https://github.com/chapel-lang/chapel/pull/20918/files) for `weakPointer`, implements the pointer as a generic record type:

```chapel
record weakPointer {
type t
...
}
```
which is defined a submodule of `sharedObject`. This type can be imported and used as follows:

```chapel
use WeakPointer;
// or import WeakPointer.weakPointer;

var myShared = new shared someClassType();

var my_weak = new weakPointer(myShared);
```

This approach is similar to Rust's `weak` and C++'s `weak_ptr`; however it differs from the other memory management strategies in the language because they are implemented using keywords rather than exposed types.

---

There are a few questions here:
- Would it be better to expose it to users as a keyword (more like `shared`, `owned`, `borrowed`)? This is more similar to Swift's interface. Or to keep it as a generic type?
- If we keep it as a type in the standard library, where should it belong? Should there be a new module for pointer types, can it go in an existing module, or is the current location ok?
- In either case, what should the type/keyword be named?

**Comment by bradcray** [comment]
Barring good reasons I'm not aware of, I'd be inclined to promote this to a class management type like `borrowed`, since it seems very similar to that.  The main downside I can think of is "reserves another keyword", but I'm not sure what the argument would be for why the other management types deserve keywords and this one doesn't.

If we were to keep it a user-facing type, things get muddier to me:  Should it be always-available, like `shared`?  (my instinct would be yes).  If so, it's effectively a built-in type, where most of our built-in types don't use camel-case and rarely refer to "pointers", so I'd be disinclined to call it `sharedPointer()`.  This also incentivizes me to use `weak` as another class qualifier since it doesn't mean I need to come up with a counterproposal for what the type would be called.

None of this needs to happen in the first draft / PR introducing the feature, but I think it should before we start proposing users use it (rather than just kicking its tires, say) so that we don't need to deprecate anything and have them update real application code.

**Comment by jeremiah-corrado** [comment]
>  but I'm not sure what the argument would be for why the other management types deserve keywords and this one doesn't

Here are a few of reasons I can think of why we might not want to promote `weak` to a the level of memory-management-strategy in the language:
1. It is sufficiently specific to `shared` that I think it ought to be more of a utility associated with `shared` alone, rather than a general purpose strategy.
2. In practice, I think it would be used much less frequently than the other strategies, so I would be hesitant to add it to the list of things that new users need to understand when they are learning about memory management in chapel. (i.e., adding `weak` to the documentation for `shared`, `owned`, `unmanaged` and `borrowed` would make it that much more intimidating to new users).
3. This might not be the last smart-pointer type that we'll want to add to the standard library. For example, I could imagine wanting a copy-on-write pointer (see Rust's [cow](https://doc.rust-lang.org/std/borrow/enum.Cow.html)) at some point. Creating a precedent that all smart-pointer types must have their own keyword in the language might inhibit the implementation of new smart-pointers. 

---

To expound on the first point, this is what the tables of valid assignments/initializations between strategies would look like with `weak` included (I think):

**Assigning y to x:** `x = y`
| x-type \ y-type| unmanaged | owned | shared | borrowed | weak |
| :-- | :-- | :-- | :-- | :-- | :-- |
| unmanaged | allowed | - | - | - | - |
| owned | - | allowed | - | - | - |
| shared | - | allowed w/ cast | allowed | - | `allowed w/ upgrade` |
| borrowed | allowed | allowed | allowed | allowed | - |
| weak | - | - | `allowed w/ downgrade` | - | allowed |

**Initializing y into x:** `x = new y`
| x-type \ y-type| unmanaged | owned | shared | borrowed | weak |
| :-- | :-- | :-- | :-- | :-- | :-- |
| unmanaged | allowed | - | - | - | - |
| owned | - | allowed | - | - | - |
| shared | - | allowed w/ cast | allowed | - | - |
| borrowed | warns | warns | warns | - | - |
| weak | - | - | - | - | - |

where the two highlighted cells show the only possible interactions between `weak` and the other MM strategies.

This makes me think that it works better as a standalone type that users would only import when they are doing something sufficiently advanced to need a weak reference to some `shared` objects.

**Comment by bradcray** [comment]
I hear what you're saying, but am not (yet) convinced.  Here are my counterarguments:

1) It's the relationship to `shared` that makes me think it should be part of the language, since `shared` is.  Specifically, since `weak` isn't being implemented as a standalone library without associated changes to the built-in `shared` type, it seems like the two are deeply intertwined, and that it would be weird to have `shared` be in the language and `weak` not.  That suggests to me that both should be in the language (or that neither should be, which would also be OK with me, though it arguably requires a solution to point 3 sooner than we'd otherwise get to it).

2) I think this, and the impact on the tables, could be mitigated by just structuring the documentation appropriately.  E.g., rather than considering it an equal sibling to the other flavors, have it be an advanced sub-topic of `shared` or presented independently of the others, don't add it to the tables, etc.

3) This resonates with me because it's something we hear from users, so I agree that it's something we need to ultimately support and have a story for.  But I'm not convinced that this is the right time to draw the line between what's built-in and what's not due to point 1.  Again, unless we decided that `shared` was not part of the language either at which point I think we'd need the ability to say something like `new <ident> C()` and have `ident` resolve to a record name that satisfied specific properties of a class memory management type.

**Comment by jeremiah-corrado** [comment]
> unless we decided that shared was not part of the language either at which point I think we'd need the ability to say something like new <ident> C() and have ident resolve to a record name that satisfied specific properties of a class memory management type.

Ok interesting, this is different than what I was imagining for other smart pointers. I was thinking that a COW, for example, would also be a generic record type that "wraps around" a class instance. Allowing for something like:
```chapel
use Cow;

proc sumAbs(numbers: [] cow(intWrapper)) {
    var s = intWrapper(0);    
    for n in numbers {
        s += if n.isPositive() then n else n.to_owned().neg();
    }
    return s;
}
```

However, if we are eventually planning on allowing custom MM strategies, then I am less hesitant to include `weak` as a lower-tier strategy. That being said, a wrapping-record type works now, without any changes to the compiler, so I'm still leaning towards the current approach. (I am not personally bothered by the inconsistency between `shared myClass`, `owned myClass` and `weak(shared myClass)`, but I think this is mostly an issue of taste. 

If we do go with a keyword, I agree with your point about documenting it such that it is clearly a specialized sub-topic of `shared` and not a sibling.

**Comment by bradcray** [comment]
@mppf: Would you help me remember:  Does `shared` itself rely on any compiler support currently (other than parsing the keyword and translating it to the implementing record, of course)?

**Comment by mppf** [comment]
> @mppf: Would you help me remember: Does `shared` itself rely on any compiler support currently (other than parsing the keyword and translating it to the implementing record, of course)?

Yes it does but that further compiler support is applied to all of the memory management records through a pragma. The compiler also relies on there being a type field with a particular name. But these things mainly have to do with supporting implicit conversions and casts on types.

One argument for why a weak pointer is in a different category is that (I think) it won't allow implicitly converting to the borrow. (Because you have to "upgrade" it first, if we want things to be relatively explicit). That means that, while we could make it syntactically similar, we probably wouldn't want the implicit conversions to borrows part to apply.

**Comment by bradcray** [comment]
> Yes it does but that further compiler support is applied to all of the memory management records through a pragma.

So if we had a general story for user-defined memory management flavors for classes, `shared` wouldn't require anything that any other user-defined memory management flavor would?  (that's what I was expecting, but wanted to be sure of).

> One argument for why a weak pointer is in a different category is that (I think) it won't allow implicitly converting to the borrow. (Because you have to "upgrade" it first, if we want things to be relatively explicit). That means that, while we could make it syntactically similar, we probably wouldn't want the implicit conversions to borrows part to apply.

To make sure I'm understanding:  Does this suggest you think that we should make it neither a syntactic sibling within the language nor one of the first user-defined memory management strategies in the modules?

**Comment by bradcray** [comment]
> arguably they are a bit more like a kind of borrow

This is definitely the way I think about it.  I agree that it's not a full memory management strategy (and don't think borrowed is either?).

> even when weak pointers are used, it's likely to be possible to write all or almost all of the code without naming the weak pointer type at all. So, a weak keyword offers relatively little benefit.

Couldn't the same be said of `borrowed`?  (or, really, many types in Chapel, like string, bytes, default numerical types, ...).

**Comment by mppf** [comment]
> > even when weak pointers are used, it's likely to be possible to write all or almost all of the code without naming the weak pointer type at all. So, a weak keyword offers relatively little benefit.
> 
> Couldn't the same be said of `borrowed`? (or, really, many types in Chapel, like string, bytes, default numerical types, ...).

I think the amount we need to use `borrowed` or `int` or `string` is much greater than the amount we would need to use `weak`.

But anyway if I take my argument to its logical conclusion we might have

``` chapel
record shared {
  record weak {
    proc getShared() { ... }
  }
  proc type weakType type { return weak(this); }
  proc getWeak(): this.type.weakType { ... }
}
```

(I'm half a mind to make it `createWeak` and `createShared` but I think the argument against that is that we tend not to think of these memory management records as a thing to be created since we don't `new shared(new MyClass)` etc.)

The point is just that the user-visible type name is `(shared MyClass).weakType` which is generally an instantiation of `shared.weak`. I think this could go fine.

Anyway, I know this is a subjective matter, but I also feel like we're trying to make a language design decision here (should `weak` be a keyword or should we use another approach?) in a vacuum because we haven't been discussing any code examples or saying. While we can argue about whether `weak MyClass` or something else is more intuitive, I think what matters is how these features come across when they are used in the context of a program.

My own opinion is that we should either:
 1. draw the line at the memory management keywords we have now; or
 2. figure out a way to allow parsing where these memory management "keywords" are defined in the library. (One way would be to try to parse <ident> <expr> as a type. Alternatively, there are some languages that have a more adaptable parsing, and we could draw from those ideas. However I think that might be a challenge to implement with the current parser).

But, maybe there is an even simpler way to move forward. Arguably the `weakType` method I sketched above in my comment is useful for convenience, even if we also have `weak` as a keyword. What if we started out with the `weakType` method being the only way to express the type, gained some experience actually using the type, and then decided whether or not we wanted to add `weak` as a keyword? (Additionally, in the future, we might be trying to fill in the details of user-defined memory management strategies; so we might figure out a way to solve 2. above at that time, for example).

**Comment by bradcray** [comment]
Thinking about this issue a bit more in the context of today's discussion about classes, I was wondering about things like:

Given:

```chapel
class C {  proc foo() { } }

class D: C {  override proc foo() { } }
```

* Do instances of `weakPointer(C)` and `weakPointer(D)` support dynamic dispatch like other flavors of `C`/`D` do?
* Can I (or should I be able to) pass a `weakPointer(C)` into a routine whose formal argument is `x: C` (of generic memory management type)?
* Can a `weakPointer(D)` have a declared static type of `weakPointer(C)`? (or should it be able to?)
* If so, can such a variable be dynamically cast to `weakPointer(D)` or `weakPointer(D?)`, similar to how `var myD: C = new D();` can be via `myD: D` or `myD: D?`?)

The more the answer to these questions is "yes", the more it feels to me like `weak C` / `weak D` is an attractive syntax for supporting this since it feels like "just another type of class memory management that behaves like the rest of them", so not deserving of a distinct type that would necessitate explanations about how it does and doesn't interact with these normal cross-memory-management capabilities of classes.

w.r.t. this comment:

> figure out a way to allow parsing where these memory management "keywords" are defined in the library. (One way would be to try to parse as a type. Alternatively, there are some languages that have a more adaptable parsing, and we could draw from those ideas. However I think that might be a challenge to implement with the current parser).

I'm more attached to writing `weak C` (or `[something] C` anyway) than I am to making `weak` reserved per se.  That said, I'd probably either make all of `owned`, `shared`, `borrowed`, `unmanaged`, and `weak` reserved or not if those are the types we're going out with (or, potentially, all of `owned`, `borrowed`, and `unmanaged` reserved and then make neither `shared` nor `weak` reserved).

> and then decided whether or not we wanted to add weak as a keyword?

(Stating the obvious): It'd be best if we could make that decision prior to 2.0.

**Comment by jeremiah-corrado** [comment]
Unfortunately I missed that meeting so I might be slightly out of the loop on this. Regardless, here are a couple of thoughts on the first two questions from above:

> - Do instances of weakPointer(C) and weakPointer(D) support dynamic dispatch like other flavors of C/D do?

I don't think so. With the current interface, calling the `foo` method on a `weakPointer(C)` or `weakPointer(D)` requires it to first be "upgraded" into a `shared C` or `shared D`, after which, I believe the dynamic dispatch rules would work as they normally do:

```chapel
var maybeD = myWeakD.upgrade();

if maybeD != nil {
  maybeD!.foo() // calls the override
  ((maybeD!): C).foo() // calls the base method
}
```

> - Can I (or should I be able to) pass a weakPointer(C) into a routine whose formal argument is x: C (of generic memory management type)

I don't think this should be allowed, mainly because I'm not sure how the generic implementation could be written to work with `weak` *and* the other strategies. I'm imagining we would need some `isWeak` method available on all classes, and then the method would need a separate branch to handle the weak case. Something along these lines:

```chapel
proc something(x: C) {
  var _x: C;
  if x.isWeak {
    var maybeX = x.upgrade();
    if maybeX != nil
      then _x = maybe_x!;
      else return;  // or throw?
  } else {
    _x = x;
  }
  // do something with '_x'
  ...
}
```

Which makes me think that it would be better to have a separate `proc something(x: weak C)` (or `proc something(x: weak(C))`) overload to handle the weak pointer case.

**Comment by bradcray** [comment]
@jeremiah-corrado's answers just above make me much less a fan of `weak C` than I had been before.  Specifically, since it seems like the type is not very class-like until/unless you manually convert it to a `shared` first, we shouldn't make it look like the other class management types.  Sorry that this distinction had been lost on me previously (particularly if people said it and I missed it).

With that, I'm much more open to representing it using an explicit class type.  Jeremiah's last example is enticing to me (`weak(C)` rather than `weakPointer(C)`).  Specifically, I like that it's shorter and doesn't bring the word 'pointer' into the mix (which, traditionally, we've tried to avoid—though with a quick look I see it's made its way into the spec with the introduction of owned/shared/borrowed classes).

**Comment by jeremiah-corrado** [comment]
During an offline meeting, we discussed a few design questions related to how a weak pointer type should be represented in the language.

Generic record vs. Memory Management strategy:
---
See the discussion above for more context on this issue. 
* generic record: 11 👍
* memory management strategy: -

Location of the weak pointer type and `shared`:
---
It was proposed that `shared` be demoted from it's status as a keyword, and instead be moved to a standard library implementation.  Under this proposal,`shared` and weakpointer would both live in this same module, which would not be included by default. There are a few reasons for this change:
- weakpointer and shared are tightly coupled (in their implementation and use) 
    - it makes sense to house them in the same module
    - Other languages do similar things: Rust's: [`rc` module](https://doc.rust-lang.org/std/rc/index.html) and [`sync` module](https://doc.rust-lang.org/std/sync/index.html) both have `Rc` and `Weak` types. C++ houses `weak_pointer` and `shared_pointer` in the `memory` namespace.
- there is a desire to eventually support custom Memory Management types (which would live in their own modules and would not need to be keywords in the language, but could still be used as follows: `var x = new myMM C()`). Putting `shared` in this category would set a precedent for other MM strategies to be implemented in the language without explicit compiler support.
- Reducing the number of default-included MM strategies from four to three simplifies the initial learning curve for `class`es in Chapel.

**Poll:**
* put shared and weak pointer together in a non-default module: 5 👍
* put weak pointer in it's own non-default module, leave `shared` as is: 1 👍
* make weak pointer available by default, leave `shared` as is: -

Not everyone was on board with this proposal, and it felt like more time was needed to discuss it before going with the leading option.

Name of the weak pointer type:
---
* `weak`: 9 👍
* `weakReference`: 1 🤔
* `weakRef`: 2 👍
* `weakShared`: 1 👍, 1 🤔

What to name a weak-pointer+shared module:
---
If we decide to put weakpointer and shared in a non-default module, what should its name be:

* `SharedObject`: 3 👍
* `SharedClass`: -
* `SharedReference`: -
* `Shared`: 8 👍

What to name a module for the weak-pointer only
---
If we decide to put weakpointer in its own non-default module, what should its name be:

* `WeakShared`: -
* `WeakReference`: -
* `WeakSharedRef`: -
* `WeakObject`: 1 👍

**Comment by bradcray** [comment]
I forked off some of the elements mentioned in Jeremiah's previous comment into an issue of their own to present the way I'm currently leaning with respect to how the features could/should be represented:  https://github.com/chapel-lang/chapel/issues/21951
